cmd/post-policy-fan-out.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2015-2023 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/s3utils"
+	"github.com/minio/minio/internal/hash"
+	xhttp "github.com/minio/minio/internal/http"
+)
+
+type multiCloser struct {
+	closers []io.Closer
+}
+
+func newMultiCloser(closers []io.Closer) io.Closer {
+	return &multiCloser{
+		closers: closers,
+	}
+}
+
+func (m *multiCloser) Close() error {
+	var err error
+	for _, c := range m.closers {
+		if e := c.Close(); e != nil {
+			err = errors.Join(err, e)
+		}
+	}
+	return err
+}
+
+// fanOutPutObject takes an input source reader and fans out multiple PUT operations
+// based on the incoming fan-out request, a context cancelation by the caller
+// would ensure all fan-out operations are canceled.
+func fanOutPutObject(ctx context.Context, bucket string, objectAPI ObjectLayer, fanOutReq []minio.PutObjectFanOutRequest, pReader *PutObjReader) ([]ObjectInfo, []error) {
```

Comment:
Sending the content as bytes makes it much easier and removes the need for multiple wrapper layers. Let me do the entire func

Reply:
This eliminates all the reader, closer, writer and does everything within goroutines:

```Go
// fanOutPutObject takes an input source reader and fans out multiple PUT operations
// based on the incoming fan-out request, a context cancelation by the caller
// would ensure all fan-out operations are canceled.
func fanOutPutObject(ctx context.Context, bucket string, objectAPI ObjectLayer, fanOutReq []minio.PutObjectFanOutRequest, content []byte) ([]ObjectInfo, []error) {
	errs := make([]error, len(fanOutReq))
	objInfos := make([]ObjectInfo, len(fanOutReq))

	var wg sync.WaitGroup
	wg.Add(len(fanOutReq))
	for i, req := range fanOutReq {
		go func(idx int, req minio.PutObjectFanOutRequest) {
			defer wg.Done()
			hr, _ := hash.NewReader(bytes.NewReader(content), int64(len(content)), "", "", -1)
			reader := NewPutObjReader(hr)
			defer func() {
				errs[idx] = errors.Join(errs[idx], reader.Close(), hr.Close())
				if errs[idx] != nil {
					objInfos[idx] = ObjectInfo{Name: fanOutReq[idx].Key}
				}
			}()
			userDefined := make(map[string]string, len(fanOutReq[idx].UserMetadata))
			for k, v := range fanOutReq[idx].UserMetadata {
				userDefined[k] = v
			}
			userDefined[xhttp.AmzObjectTagging] = s3utils.TagEncode(fanOutReq[idx].UserTags)

			objInfos[idx] , errs[idx] = objectAPI.PutObject(ctx, bucket, fanOutReq[idx].Key, reader, ObjectOptions{
				Versioned:        globalBucketVersioningSys.PrefixEnabled(bucket, fanOutReq[idx].Key),
				VersionSuspended: globalBucketVersioningSys.PrefixSuspended(bucket, fanOutReq[idx].Key),
				UserDefined:      userDefined,
			})
		}(i, req)
	}
	wg.Wait()

	return objInfos, errs
}
```

Reply:
Well, there is a reason we need to do that because of the encryption requirement @klauspost that's why I can't pass down bytes as plain-text. 

If SSE headers are set we need a stream that is precisely the encrypted stream.

Reply:
@harshavardhana Before calling you can read the stream. You are not mutating any data per upload anyway.

If data should have unique keys we need to move the encryption part down here, so targets can have unique keys. Either way sending data as bytes will allow writes to complete independent of each other and not be tied to a single io.Copy loop.